From 8dcbb2d4ca5a1121225d87150414b836d3b964a2 Mon Sep 17 00:00:00 2001
From: Tim Lunn <tim@feathertop.org>
Date: Sun, 1 Mar 2015 09:24:26 +1100
Subject: [PATCH 1/3] Use traditional titlebars on Unity

For environmnets with more traditional UI, use headerbar as a toolbar and
keep traditional titlebars
---
 data/ui/contacts-window.ui | 206 ++++++++++++++++++++++++---------------------
 src/contacts-window.vala   |  33 +++++++-
 2 files changed, 138 insertions(+), 101 deletions(-)

diff --git a/data/ui/contacts-window.ui b/data/ui/contacts-window.ui
index 5272ecd..00d8973 100644
--- a/data/ui/contacts-window.ui
+++ b/data/ui/contacts-window.ui
@@ -13,140 +13,150 @@
       <placeholder />
     </child>
     <child>
-      <object class="GtkStack" id="view_switcher">
+      <object class="GtkBox" id="vbox">
+        <property name="orientation">vertical</property>
         <property name="visible">True</property>
-        <property name="can_focus">False</property>
         <child>
-          <object class="GtkOverlay" id="overlay">
+          <object class="GtkStack" id="view_switcher">
             <property name="visible">True</property>
             <property name="can_focus">False</property>
             <child>
-              <object class="GtkGrid" id="content_grid">
+              <object class="GtkOverlay" id="overlay">
                 <property name="visible">True</property>
                 <property name="can_focus">False</property>
                 <child>
-                  <object class="GtkOverlay">
+                  <object class="GtkGrid" id="content_grid">
                     <property name="visible">True</property>
                     <property name="can_focus">False</property>
                     <child>
-                      <object class="ContactsContactPane" id="contact_pane">
-                        <property name="show_tabs">False</property>
+                      <object class="GtkOverlay">
                         <property name="visible">True</property>
-                        <property name="hexpand">True</property>
-                        <signal name="will-delete" handler="contact_pane_delete_contact_cb" object="ContactsWindow" after="no" swapped="no"/>
-                        <signal name="contacts-linked" handler="contact_pane_contacts_linked_cb" object="ContactsWindow" after="no" swapped="no"/>
+                        <property name="can_focus">False</property>
+                        <child>
+                          <object class="ContactsContactPane" id="contact_pane">
+                            <property name="show_tabs">False</property>
+                            <property name="visible">True</property>
+                            <property name="hexpand">True</property>
+                            <signal name="will-delete" handler="contact_pane_delete_contact_cb" object="ContactsWindow" after="no" swapped="no"/>
+                            <signal name="contacts-linked" handler="contact_pane_contacts_linked_cb" object="ContactsWindow" after="no" swapped="no"/>
+                          </object>
+                        </child>
                       </object>
+                      <packing>
+                        <property name="left_attach">1</property>
+                        <property name="top_attach">0</property>
+                        <property name="width">1</property>
+                        <property name="height">1</property>
+                      </packing>
+                    </child>
+                    <child>
+    		  <object class="GtkGrid" id="loading_box">
+    		    <property name="visible">True</property>
+    		    <property name="can_focus">False</property>
+    		    <property name="hexpand">True</property>
+    		    <property name="vexpand">True</property>
+    		    <property name="orientation">vertical</property>
+    		    <property name="width_request">300</property>
+    		    <property name="row_homogeneous">True</property>
+    		    <child>
+    		      <object class="GtkSpinner">
+    			<property name="visible">True</property>
+    			<property name="can_focus">False</property>
+    			<property name="active">True</property>
+    			<property name="valign">end</property>
+    			<property name="halign">center</property>
+    			<style>
+    			  <class name="contacts-watermark"/>
+    			</style>
+    		      </object>
+    		      <packing>
+    			<property name="left_attach">0</property>
+    			<property name="top_attach">0</property>
+    			<property name="width">1</property>
+    			<property name="height">1</property>
+    		      </packing>
+    		    </child>
+    		    <child>
+    		      <object class="GtkLabel" id="label1">
+    			<property name="visible">True</property>
+    			<property name="can_focus">False</property>
+    			<property name="vexpand">True</property>
+    			<property name="valign">start</property>
+    			<property name="hexpand">True</property>
+    			<property name="halign">center</property>
+    			<property name="label" translatable="yes">Loading</property>
+    			<style>
+    			  <class name="contacts-watermark"/>
+    			</style>
+    		      </object>
+    		      <packing>
+    			<property name="left_attach">0</property>
+    			<property name="top_attach">1</property>
+    			<property name="width">1</property>
+    			<property name="height">1</property>
+    		      </packing>
+    		    </child>
+    		  </object>
+                      <packing>
+                        <property name="left_attach">0</property>
+                        <property name="top_attach">0</property>
+                        <property name="width">1</property>
+                        <property name="height">1</property>
+                      </packing>
                     </child>
                   </object>
+                </child>
+              </object>
+              <packing>
+                <property name="name">content-view</property>
+              </packing>
+            </child>
+            <child>
+              <object class="GtkGrid">
+                <property name="visible">True</property>
+                <property name="border_width">12</property>
+                <property name="row_spacing">24</property>
+                <property name="orientation">vertical</property>
+                <child>
+                  <object class="GtkLabel" id="setup_explanation_label">
+                    <property name="visible">True</property>
+                    <property name="halign">center</property>
+                    <property name="wrap">True</property>
+                    <property name="max_width_chars">25</property>
+                    <property name="label">New contacts will be added to the selected address book. You are able to view and edit contacts from other address books.</property>
+                  </object>
                   <packing>
-                    <property name="left_attach">1</property>
+                    <property name="left_attach">0</property>
                     <property name="top_attach">0</property>
                     <property name="width">1</property>
                     <property name="height">1</property>
                   </packing>
                 </child>
                 <child>
-		  <object class="GtkGrid" id="loading_box">
-		    <property name="visible">True</property>
-		    <property name="can_focus">False</property>
-		    <property name="hexpand">True</property>
-		    <property name="vexpand">True</property>
-		    <property name="orientation">vertical</property>
-		    <property name="width_request">300</property>
-		    <property name="row_homogeneous">True</property>
-		    <child>
-		      <object class="GtkSpinner">
-			<property name="visible">True</property>
-			<property name="can_focus">False</property>
-			<property name="active">True</property>
-			<property name="valign">end</property>
-			<property name="halign">center</property>
-			<style>
-			  <class name="contacts-watermark"/>
-			</style>
-		      </object>
-		      <packing>
-			<property name="left_attach">0</property>
-			<property name="top_attach">0</property>
-			<property name="width">1</property>
-			<property name="height">1</property>
-		      </packing>
-		    </child>
-		    <child>
-		      <object class="GtkLabel" id="label1">
-			<property name="visible">True</property>
-			<property name="can_focus">False</property>
-			<property name="vexpand">True</property>
-			<property name="valign">start</property>
-			<property name="hexpand">True</property>
-			<property name="halign">center</property>
-			<property name="label" translatable="yes">Loading</property>
-			<style>
-			  <class name="contacts-watermark"/>
-			</style>
-		      </object>
-		      <packing>
-			<property name="left_attach">0</property>
-			<property name="top_attach">1</property>
-			<property name="width">1</property>
-			<property name="height">1</property>
-		      </packing>
-		    </child>
-		  </object>
+                  <object class="ContactsAccountsList" id="setup_accounts_list">
+                    <property name="visible">True</property>
+                    <property name="hexpand">True</property>
+                    <property name="halign">center</property>
+                  </object>
                   <packing>
                     <property name="left_attach">0</property>
-                    <property name="top_attach">0</property>
+                    <property name="top_attach">1</property>
                     <property name="width">1</property>
                     <property name="height">1</property>
                   </packing>
                 </child>
               </object>
-            </child>
-          </object>
-          <packing>
-            <property name="name">content-view</property>
-          </packing>
-        </child>
-        <child>
-          <object class="GtkGrid">
-            <property name="visible">True</property>
-            <property name="border_width">12</property>
-            <property name="row_spacing">24</property>
-            <property name="orientation">vertical</property>
-            <child>
-              <object class="GtkLabel" id="setup_explanation_label">
-                <property name="visible">True</property>
-                <property name="halign">center</property>
-                <property name="wrap">True</property>
-                <property name="max_width_chars">25</property>
-                <property name="label">New contacts will be added to the selected address book. You are able to view and edit contacts from other address books.</property>
-              </object>
               <packing>
-                <property name="left_attach">0</property>
-                <property name="top_attach">0</property>
-                <property name="width">1</property>
-                <property name="height">1</property>
-              </packing>
-            </child>
-            <child>
-              <object class="ContactsAccountsList" id="setup_accounts_list">
-                <property name="visible">True</property>
-                <property name="hexpand">True</property>
-                <property name="halign">center</property>
-              </object>
-              <packing>
-                <property name="left_attach">0</property>
-                <property name="top_attach">1</property>
-                <property name="width">1</property>
-                <property name="height">1</property>
+                <property name="name">setup-view</property>
               </packing>
             </child>
           </object>
           <packing>
-            <property name="name">setup-view</property>
+            <property name="pack_type">end</property>
           </packing>
         </child>
-      </object>
+        </object>
+
     </child>
   </template>
   <object class="GtkSizeGroup">
diff --git a/src/contacts-window.vala b/src/contacts-window.vala
index c49d6a7..940f0c6 100644
--- a/src/contacts-window.vala
+++ b/src/contacts-window.vala
@@ -29,6 +29,8 @@ public class Contacts.Window : Gtk.ApplicationWindow {
   [GtkChild]
   private SizeGroup left_pane_size_group;
   [GtkChild]
+  private Box vbox;
+  [GtkChild]
   private HeaderBar left_toolbar;
   [GtkChild]
   private HeaderBar right_toolbar;
@@ -141,7 +143,7 @@ public class Contacts.Window : Gtk.ApplicationWindow {
 
     if ((app as App).settings.get_boolean ("did-initial-setup")) {
       view_switcher.visible_child_name = "content-view";
-      set_titlebar (content_header_bar);
+      set_headerbar (content_header_bar);
     } else {
       var change_book_action = app.lookup_action ("change_book") as GLib.SimpleAction;
       if (change_book_action != null) {
@@ -158,13 +160,13 @@ public class Contacts.Window : Gtk.ApplicationWindow {
 	});
 
       view_switcher.visible_child_name = "setup-view";
-      set_titlebar (setup_header_bar);
+      set_headerbar (setup_header_bar);
 
       setup_accounts_list.update_contents (false);
 
       setup_done_button.clicked.connect (() => {
 	  view_switcher.visible_child_name = "content-view";
-	  set_titlebar (content_header_bar);
+	  set_headerbar (content_header_bar);
 
 	  var e_store = setup_accounts_list.selected_store as Edsf.PersonaStore;
 	  eds_source_registry.set_default_address_book (e_store.source);
@@ -182,6 +184,31 @@ public class Contacts.Window : Gtk.ApplicationWindow {
     init_content_widgets ();
   }
 
+  private void set_headerbar (Widget header_bar) {
+    var desktop = Environment.get_variable ("XDG_CURRENT_DESKTOP");
+
+    if (desktop == null || desktop.contains("GNOME"))
+    {
+      this.set_titlebar (header_bar);
+    }
+    else
+    {
+      this.set_title (_("Contacts"));
+      if (view_switcher.visible_child_name == "setup-view")
+      {
+        setup_header_bar.get_style_context().remove_class("titlebar");
+      }
+      else
+      {
+        left_toolbar.show_close_button = false;
+        right_toolbar.show_close_button = false;
+        left_toolbar.get_style_context().remove_class("titlebar");
+        right_toolbar.get_style_context().remove_class("titlebar");
+        setup_header_bar.destroy();
+      }
+      vbox.pack_start(header_bar,  false, false, 0);
+    }
+  }
   public void set_list_pane () {
     /* FIXME: if no contact is loaded per backend, I must place a sign
      * saying "import your contacts/add online account" */
-- 
2.1.4

